src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSource.cs

```diff
@@ -0,0 +1,171 @@
+            {
+                vts.SetException(error);
+            }
+            else
+            {
+                vts.SetResult(result);
+            }
+            return vts;
+        }
+
+        public static ManualResetValueTaskSource<T> Delay<T>(int delayMs, T result, Exception error = null)
+        {
+            var vts = new ManualResetValueTaskSource<T>();
+            Task.Delay(delayMs).ContinueWith(_ =>
+            {
+                if (error != null)
+                {
+                    vts.SetException(error);
+                }
+                else
+                {
+                    vts.SetResult(result);
+                }
+            });
+            return vts;
+        }
+    }
+
+    internal sealed class ManualResetValueTaskSource<T> : IValueTaskSource<T>, IValueTaskSource
+    {
+        private static readonly Action<object> s_sentinel = new Action<object>(s => { });
+        private Action<object> _continuation;
+        private object _continuationState;
+        private SynchronizationContext _capturedContext;
+        private ExecutionContext _executionContext;
+        private bool _completed;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+
+        public ValueTaskSourceStatus GetStatus(short token) =>
+            !_completed ? ValueTaskSourceStatus.Pending :
+            _error == null ? ValueTaskSourceStatus.Succeeded :
+            _error.SourceException is OperationCanceledException ? ValueTaskSourceStatus.Canceled :
+            ValueTaskSourceStatus.Faulted;
+
+        public T GetResult(short token)
+        {
+            if (!_completed)
+            {
+                throw new Exception("Not completed");
+            }
+
+            ExecutionContext ctx = _executionContext;
+            if (ctx != null)
+            {
+                ctx.Dispose();
+                _executionContext = null;
+            }
+
+            _error?.Throw();
+            return _result;
+        }
+
+        void IValueTaskSource.GetResult(short token)
+        {
+            GetResult(token);
+        }
+
+        public void Reset()
+        {
+            _completed = false;
+            _continuation = null;
+            _continuationState = null;
+            _result = default;
+            _error = null;
+        }
+
+        public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            if ((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) != 0)
+            {
+                _executionContext = ExecutionContext.Capture();
+            }
+
+            if ((flags & ValueTaskSourceOnCompletedFlags.UseSchedulingContext) != 0)
+            {
+                _capturedContext = SynchronizationContext.Current;
+            }
+
+            _continuationState = state;
+            if (Interlocked.CompareExchange(ref _continuation, continuation, null) != null)
```

Comment:
Throw on failure?

(I realize it's just a test class, but it seems useful as sample code as well...)